scripts/runnable_recipe_ray_vllm_weight_sync.py

```diff
@@ -368,6 +369,447 @@ def run(self):
+        else:
+            policy_version = 0
+
+        response_padding_masks = torch.eq(response_tokens, self._tokenizer.pad_id)
+        seq_lens = training.get_unmasked_sequence_lengths(response_padding_masks)
+        del response_padding_masks
+
+        postprocessed_results = (
+            query_responses,
+            response_tokens,
+            logprobs,
+            query_response_padding_masks,
+            seq_lens,
+            answers,
+            policy_version
+        )
+        return postprocessed_results
+
+
+    def run(self):
+        for i in range(63):
+            self.rollout()
+            print(f"{self.worker_id} about to update weights, {self.remote_weight_updater}")
+            if i > 1:
+                self.remote_weight_updater.update_weights.remote(weights=None, worker_ids=self.worker_id)
+                self.local_weight_updater()
+
+
+    def rollout(self) -> TensorDictBase:
+        if self.reset_at_each_iter or self._shuttle is None:
+            data = self.env.reset()
+        else:
+            data = self._shuttle
+
+        tensordicts = []
+        collected_frames = 0
+        while collected_frames < self.frames_per_batch:
+            policy_input = data
+            env_input = self.policy(
+                            policy_input,
+                            self.llm,
+                            **self.policy_kwargs,
+                        )
+            env_output, env_next_output = self.env.step_and_maybe_reset(env_input)
+
+            # carry over collector data without messing up devices
+            # collector_data = env_output.get("collector").copy()
+            collector_data = self._shuttle.get("collector").copy()
+            env_next_output.set("collector", collector_data)
+            self._shuttle = env_next_output
+            if self._shuttle_has_no_device:
+                self._shuttle.clear_device_()
+            self._shuttle.set("collector", collector_data)
+            self._update_traj_ids(env_output)
+            data = self._shuttle
+            tensordicts.append(data)
+            collected_frames += data.numel()
+
+        data = lazy_stack(tensordicts, -1)
+
+        if self.replay_buffer is not None:
+            self.replay_buffer.extend(data)
+            return None
+        
+        if self.rollout_queue is not None:
+            assert self.replay_buffer is None
+            postprocessed_results = self._postprocess_for_queue(data)
+            print(postprocessed_results)
+
+            while True:
+                try:
+                    self.rollout_queue.put_nowait(postprocessed_results)
+                    break
+                except QueueFull:
+                    self.rollout_queue.get()  # Remove the oldest item to make space
+                    print(f"rollout queue full. Discarding data.")
+
+        return data
+    
+    def _setup_data(
```

Comment:
This is fine for now

Reply:
This is almost directly from the tune script

Reply:
There was a TODO comment, just saying it's not urgent at all!